com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs

```diff
@@ -78,21 +212,88 @@ public int Add(in T value)
         /// <param name="keepContent">Set to true if you want the current content of the array to be kept.</param>
         public void Resize(int newSize, bool keepContent = false)
         {
-            if (newSize > m_Array.Length)
+            Reserve(newSize, keepContent);
+            size = newSize;
+        }
+
+        /// <summary>
+        /// Sets the total number of elements the internal data structure can hold without resizing.
+        /// </summary>
+        /// <param name="newCapacity">New capacity for the array.</param>
+        /// <param name="keepContent">Set to true if you want the current content of the array to be kept.</param>
+        public void Reserve(int newCapacity, bool keepContent = false)
+        {
+            if (newCapacity > m_Array.Length)
             {
                 if (keepContent)
                 {
-                    var newArray = new T[newSize];
+                    var newArray = new T[newCapacity];
                     Array.Copy(m_Array, newArray, m_Array.Length);
                     m_Array = newArray;
                 }
                 else
                 {
-                    m_Array = new T[newSize];
+                    m_Array = new T[newCapacity];
                 }
             }
+        }
 
-            size = newSize;
+        static int Partition<U>(U[] data, int left, int right) where U : IComparable<U>, new()
+        {
+            var pivot = data[left];
+
+            --left;
+            ++right;
+            while (true)
+            {
+                var c = 0;
+                var lvalue = default(U);
+                do
+                {
+                    ++left;
+                    lvalue = data[left];
+                    c = lvalue.CompareTo(pivot);
+                }
+                while (c < 0);
+
+                var rvalue = default(U);
+                do
+                {
+                    --right;
+                    rvalue = data[right];
+                    c = rvalue.CompareTo(pivot);
+                }
+                while (c > 0);
+
+                if (left < right)
+                {
+                    data[right] = lvalue;
+                    data[left] = rvalue;
+                }
+                else
+                {
+                    return right;
+                }
+            }
+        }
+
+        static void QuickSort<U>(U[] data, int left, int right) where U : IComparable<U>, new()
+        {
+            if (left < right)
+            {
+                int pivot = Partition(data, left, right);
+
+                if (pivot >= 1)
+                    QuickSort(data, left, pivot);
+
+                if (pivot + 1 < right)
+                    QuickSort(data, pivot + 1, right);
+            }
+        }
+
+        public static void QuickSort<U>(DynamicArray<U> array) where U : IComparable<U>, new()
```

Comment:
You should be able to achieve better API by making this an [extension method](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods). For that, it's needs to be declared in a separate static non-nested class, naming convention here would be `public static class DynamicArrayExtensions`. Signature something like this: `public static void QuickSort<T>(this DynamicArray<T> array) where T : IComparable<T>, new()`. That way you can keep the stricter type requirements for this function only, but calling it looks like just calling a normal method: `m_DynamicArray.QuickSort();`

Reply:
I did not think of that. I knew it was a good idea to add you as a reviewer :)
Will try to implement that.

Reply:
Updated the PR with the extension. Looks much cleaner now.